src/algebra/order/lattice_group.lean

```diff
@@ -460,3 +460,23 @@ begin
 end
 
 end lattice_ordered_comm_group
+
+namespace lattice_ordered_add_comm_group
+
+variables {β : Type u} [lattice β] [add_comm_group β]
+
+section solid
+
+/-- A subset `s ⊆ β`, with `β` an `add_comm_group` with a `lattice` structure, is solid if for
+all `x ∈ s` and all `y ∈ β` such that `|y| ≤ |x|`, then `y ∈ s`. -/
+def is_solid (s : set β) : Prop := ∀ ⦃x⦄, x ∈ s → ∀ ⦃y⦄, |y| ≤ |x| → y ∈ s
+
+/-- The solid closure of a subset `s` is the smallest superset of `s` that is solid. -/
```

Comment:
We show below that the solid closure is solid. I think we're missing the proof that it's the smallest solid set containing `s`?

Reply:
That should be called `solid_closure_min`, cf `convex_hull_min` for the arguments order.

Reply:
As an aside, Isabelle / HOL abstracts out the concept of hull, which occurs in several places: https://isabelle.in.tum.de/library/HOL/HOL/Hull.html

Reply:
> As an aside, Isabelle / HOL abstracts out the concept of hull, which occurs in several places: https://isabelle.in.tum.de/library/HOL/HOL/Hull.html

I guess `order.closure` is the equivalent in Mathlib?

Reply:
Would it make sense to use the closure API here? E.g. something like:
```
def solid_closure' : closure_operator (set β) := closure_operator.mk'
  (λ s, {y : β | ∃ x ∈ s, |y| ≤ |x|} )
  (λ s t hst, begin
    simp only [exists_prop, set.le_eq_subset, set.set_of_subset_set_of, forall_exists_index, and_imp],
    intros x y z hxy,
    use y,
    split,
    { exact hst z, },
    { exact hxy, }
    end)
  (λ s, begin
    simp only [exists_prop, set.le_eq_subset],
    intros x hx,
    rw set.mem_set_of_eq,
    use x,
    split,
    { exact hx, },
    { exact rfl.ge, }
  end)
  (λ s, begin
    simp only [exists_prop, set.mem_set_of_eq, set.le_eq_subset, set.set_of_subset_set_of,
      forall_exists_index, and_imp],
      intros x y z hz hyz hxy,
      use z,
      split,
      { exact hz, },
      { exact le_trans hxy hyz, }
  end)
```

Reply:
Or, for closer alignment with hull,
```lean
lemma is_solid_sInter {S : set (set β)} (h : ∀ s ∈ S, is_solid s) : is_solid (⋂₀ S) :=
λ y hy z hyz s hs, h s hs (hy s hs) hyz

lemma is_solid_Inter {ι : Sort*} {s : ι → set β} (h : ∀ i, is_solid (s i)) : is_solid (⋂ i, s i) :=
  (set.sInter_range s) ▸ is_solid_sInter $ set.forall_range_iff.2 h

def solid_closure'' : closure_operator (set β) := closure_operator.mk₃
  (λ s, ⋂ (t : set β) (hst : s ⊆ t) (ht : is_solid t), t)
  (is_solid)
  (λ s, set.subset_Inter (λ t, set.subset_Inter $ λ hst, set.subset_Inter $ λ ht, hst))
  (λ s, is_solid_Inter $ λ t, is_solid_Inter $ λ ht, is_solid_Inter id)
  (λ s t hst ht, set.Inter_subset_of_subset t $ set.Inter_subset_of_subset hst $
  set.Inter_subset _ ht)
```

Reply:
Please don't use `closure_operator`. I botched it two years ago.

Reply:
> Please don't use `closure_operator`. I botched it two years ago.

So it should be implemented as a Galois insertion instead?

Reply:
Something like this?
```lean
def adjoin (s : set β) : { t : set β // is_solid t} := ⟨{y | ∃ x ∈ s, |y| ≤ |x|}, λ x ⟨y, hy, hxy⟩ z hzx, ⟨y, hy, hzx.trans hxy⟩⟩  

protected lemma gc : galois_connection (adjoin : set β → { t : set β // is_solid t}) coe := sorry

protected def gi : galois_insertion (adjoin : set β → { t : set β // is_solid t}) coe := sorry
```

?

Reply:
I think we definitely need the result that says that `solid_closure` is indeed the solid closure. I'll add it. Besides that, I guess we should only add some additional results if we have use for them. 

I'd like to get some feedback on my question from above: 

> At the moment, is_solid_ball lives in the root namespace. I could move it to the lattice_ordered_add_comm_group (with is_solid), but does that makes sense when we have [dual_solid](https://leanprover-community.github.io/mathlib_docs/analysis/normed/order/lattice.html#dual_solid) living also in the root namespace? And in fact maybe is_solid_ball should be proved only for normed_lattice_add_comm_group?

Reply:
I don't have an overview of this field. So, let's be conservative for now, and namespace things. We can always put things back in the root namespace later.